test/stdlib/Result.swift

```diff
@@ -0,0 +1,155 @@
+
+import StdlibUnittest
+import Swift
+
+let ResultTests = TestSuite("Result")
+
+fileprivate enum Err: Error, Equatable {
+  case err
+  case derr
+}
+
+fileprivate let string = "string"
+
+ResultTests.test("Construction") {
+  let result1: Result<String, String> = .success(string)
+  let result2: Result<String, String> = .failure(string)
+  let string1: String? = {
+    switch result1 {
+      case let .success(string): 
+        return string
+      case .failure: 
+        return nil
+    }
+  }()
+  let string2: String? = {
+    switch result2 {
+      case let .failure(string): 
+        return string
+      case .success: 
+        return nil
+    }
+  }()
+
+  expectEqual(string1, string2)
+}
+
+ResultTests.test("Properties") {
+  let result1: Result<String, String> = .success(string)
+  let result2: Result<String, String> = .failure(string)
+
+  expectEqual(result1.value, result2.error)
+  expectTrue(result1.isSuccess)
+  expectFalse(result2.isSuccess)
+}
+
+ResultTests.test("Throwing Initialization and Unwrapping") {
+  func notThrowing() throws -> String {
+    return string
+  }
+
+  func throwing() throws -> String {
+    throw Err.err
+  }
+    
+  let result1 = Result { try throwing() }
+  let result2 = Result { try notThrowing() }
+    
+  expectEqual(result1.error as? Err, Err.err)
+  expectEqual(result2.value, string)
+    
+  do {
+    _ = try result1.unwrapped()
+  } catch let error as Err {
+    expectEqual(error, Err.err)
+  } catch {
+    preconditionFailure("unexpected error type")
+  }
+    
+  do {
+    let unwrapped = try result2.unwrapped()
+    expectEqual(unwrapped, string)
+  } catch {
+    preconditionFailure("unexpected error")
+  }
+    
+  // Test unwrapping strongly typed error.
+  let result3 = Result<String, Err>.failure(Err.err)
+  do {
+    _ = try result3.unwrapped()
+  } catch let error as Err {
+    expectEqual(error, Err.err)
+  } catch {
+    preconditionFailure("unexpected error type")
+  }
+}
+
+ResultTests.test("Functional Transforms") {
+  func transform(_ int: Int) -> Int {
+    return 2 * int
+  }
+
+  func resultTransform(_ int: Int) -> Result<Int, Int> {
+    return .success(transform(int))
+  }
+  
+  func throwingTransform(_ int: Int) throws -> Int {
+    throw Err.err
+  }
+    
+  let result1: Result<Int, Int> = .success(1)
+  let newResult1 = result1.map(transform)
+    
+  expectEqual(newResult1, .success(2))
+    
+  let result2: Result<Int, Int> = .failure(1)
+  let newResult2 = result2.mapError(transform)
+    
+  expectEqual(newResult2, .failure(2))
+    
+  let result3: Result<Int, Int> = .success(1)
+  let newResult3 = result3.flatMap(resultTransform)
+    
+  expectEqual(newResult3, .success(2))
+    
+  let result4: Result<Int, Int> = .failure(1)
+  let newResult4 = result4.flatMapError(resultTransform)
+    
+  expectEqual(newResult4, .success(2))
+    
+  let result5: Result<Int, Int> = .success(1)
+  let output1 = result5.fold(onSuccess: transform, onFailure: transform)
```

Comment:
Why not prove that `fold(onSuccess:onFailure:)` called the proper closure by passing a different transform for the other one?

Reply:
Added a different closure for the failure case.